ipa-core/src/test_fixture/world.rs

```diff
@@ -961,4 +1004,65 @@ mod tests {
             assert_eq!(shares[1].right(), shares[2].left());
         });
     }
+
+    #[test]
+    fn zkp_malicious_sharded() {
+        run(|| async {
+            let world: TestWorld<WithShards<2>> =
+                TestWorld::with_shards(TestWorldConfig::default());
+            let input = vec![Boolean::truncate_from(0_u32), Boolean::truncate_from(1_u32)];
+            let r = world
+                .malicious(input.clone().into_iter(), |ctx, input| async move {
+                    assert_eq!(1, input.iter().len());
+                    let ctx = ctx.set_total_records(TotalRecords::ONE);
+                    let validator = ctx.dzkp_validator(TEST_DZKP_STEPS, 1);
+                    let ctx = validator.context();
+                    let r = input[0]
+                        .multiply(&input[0], ctx, RecordId::FIRST)
+                        .await
+                        .unwrap();
+                    validator.validate().await.unwrap();
+
+                    vec![r]
+                })
+                .await
+                .into_iter()
+                .flat_map(|v| v.reconstruct())
+                .collect::<Vec<_>>();
+
+            assert_eq!(input, r);
+        });
+    }
+
+    #[test]
+    fn mac_malicious_sharded() {
+        run(|| async {
+            let world: TestWorld<WithShards<2>> =
+                TestWorld::with_shards(TestWorldConfig::default());
+            let input = vec![Fp31::truncate_from(0_u32), Fp31::truncate_from(1_u32)];
+            let r = world
+                .malicious(input.clone().into_iter(), |ctx, input| async move {
+                    assert_eq!(1, input.iter().len());
+                    let validator = ctx.set_total_records(1).validator();
+                    let ctx = validator.context();
+                    let (a_upgraded, b_upgraded) = (input[0].clone(), input[0].clone())
+                        .upgrade(ctx.clone(), RecordId::FIRST)
+                        .await
+                        .unwrap();
+                    let _ = a_upgraded
```

Comment:
Wait a second, why aren't you using the result of the multiplication? Surely you meant to return this, not `input`, because then it's trivially true that input == input.

Reply:
probably not in this case because these shares are upgraded and you can't reconstruct from them. It is possible to do downgrade and then return, but the purpose of this test is to check that `validate_record` and `upgrade` API are available inside sharded infrastructure, so the output does not really matter here.